dialer.go

```diff
@@ -68,175 +81,177 @@ func NewDialer(opts *DialerOpts) Dialer {
+	s, err := d.getSession(ctx)
+	go d.recycleSession(s)
+	if err != nil {
+		return nil, err
+	}
+	return s.createStream(ctx), nil
+}
+
+func (d *dialer) getSession(ctx context.Context) (sessionIntf, error) {
+	d.sessionRequests <- true
+	start := time.Now()
 	for {
 		select {
 		case s := <-d.liveSessions:
-			if s.AllowNewStream(d.maxStreamsPerConn, d.idleInterval) {
-				return s, nil
+			if s.isClosed() {
+				// Closed sessions will trigger the creation of new ones, so keep waiting for a new session.
+				continue
 			}
-			d.muNumLivePending.Lock()
-			d.numLive--
-			d.muNumLivePending.Unlock()
-			s.MarkDefunct()
-			newSession(minLiveConns)
-		case <-time.After(d.redialSessionInterval):
-			newSession(d.maxLiveConns)
+			log.Debugf("Returning session in %v", time.Since(start))
+			return s, nil
+
 		case <-ctx.Done():
-			return nil, errors.New("No session available")
+			err := fmt.Errorf("no session available after %v to %v", time.Since(start), d.name)
+			return nil, err
 		}
 	}
 }
 
-func (d *dialer) returnSession(s sessionIntf) {
-	addBack := true
-	d.muNumLivePending.Lock()
-	if d.numLive > minLiveConns {
-		d.numLive--
-		addBack = false
-	}
-	d.muNumLivePending.Unlock()
-	if addBack {
-		d.liveSessions <- s
-	} else {
-		s.MarkDefunct()
-	}
-}
-
 func (d *dialer) EMARTT() time.Duration {
 	return d.emaRTT.GetDuration()
 }
 
-func (d *dialer) BoundTo(dial DialFN) BoundDialer {
-	return &boundDialer{d, dial}
-}
-
-func (d *dialer) startSession(dial DialFN) (*session, error) {
-	conn, err := dial()
-	if err != nil {
-		return nil, err
-	}
-
+func (d *dialer) startSession(rs *sessionConfig) (*session, error) {
+	lc := d.lifecyle.OnTCPStart()
 	cs, err := newCryptoSpec(d.cipherCode)
 	if err != nil {
-		return nil, fmt.Errorf("Unable to create crypto spec for %v: %v", d.cipherCode, err)
+		lc.OnSessionError(err, err)
+		return nil, fmt.Errorf("unable to create crypto spec for %v: %v", d.cipherCode, err)
 	}
 
 	// Generate the client init message
 	clientInitMsg, err := buildClientInitMsg(d.serverPublicKey, d.windowSize, d.maxPadding, cs)
 	if err != nil {
-		return nil, fmt.Errorf("Unable to generate client init message: %v", err)
+		lc.OnSessionError(err, err)
+		return nil, fmt.Errorf("unable to generate client init message: %v", err)
 	}
 
-	return startSession(conn, d.windowSize, d.maxPadding, false, d.pingInterval, cs, clientInitMsg, d.pool, d.emaRTT, nil, nil)
-}
-
-type boundDialer struct {
-	Dialer
-
-	dial DialFN
-}
+	conn, err := d.dial(rs.dialTimeout)
+	if err != nil {
+		lc.OnTCPConnectionError(err)
+		return nil, err
+	}
+	lc.OnTCPEstablished(conn)
 
-func (bd *boundDialer) Dial() (net.Conn, error) {
-	return bd.Dialer.Dial(bd.dial)
-}
+	s, err := startSession(conn, d.windowSize, d.maxPadding, false, d.pingInterval, cs, clientInitMsg, d.pool,
+		d.emaRTT, nil, nil, rs, lc)
 
-func (bd *boundDialer) DialContext(ctx context.Context) (net.Conn, error) {
-	return bd.Dialer.DialContext(ctx, bd.dial)
+	if err != nil {
```

Comment:
Should close conn if unable to start session. Again, not super likely but better to be clean about it.

Reply:
Yup good catch fixed!